### PR TITLE
Skip tiles with no data

### PR DESCRIPTION
**Pull request by sgillies**
We detect tiles with no valid data by checking a mask of an in-memory temp dataset.

### CODE REVIEW

File: mbtiles/__init__.py [hunk 1]
```diff
@@ -47,19 +50,35 @@ def process_tile(tile):
...
+    kwds['driver'] = 'GTiff'
+
+    # Reproject into a GeoTIFF tile.
+    with MemoryFile() as memdst:
+        with memdst.open(**kwds) as tiledst:
+            reproject(
+                rasterio.band(src, src.indexes),
+                rasterio.band(tiledst, tiledst.indexes),
+                src_nodata=src_nodata, dst_nodata=dst_nodata, num_threads=2,
+                resampling=resampling)
```
Comment (sgillies) [inline mbtiles/__init__.py]: Switching to `MemoryFile` requires rasterio>=1.0a6.

File: mbtiles/__init__.py [hunk 2]
```diff
@@ -47,19 +50,35 @@ def process_tile(tile):
...
+        # To save time, we check the only mask of the first band for
+        # valid data.
+        with memdst.open() as tiledst:
+            mask = tiledst.read_masks(1)
+            if mask.any():
+                # We're using MemoryFile to get a filename from
+                # the in-memory filesystem only; it doesn't support
+                # copy yet.
+                with MemoryFile() as memimg:
+                    copy(tiledst, memimg.name, driver=base_kwds['driver'])
```
Comment (sgillies) [inline mbtiles/__init__.py]: This new version of copy requires rasterio>=1.0a11

File: mbtiles/__init__.py [hunk 3]
```diff
@@ -47,19 +50,35 @@ def process_tile(tile):
...
+                rasterio.band(tiledst, tiledst.indexes),
+                src_nodata=src_nodata, dst_nodata=dst_nodata, num_threads=2,
+                resampling=resampling)
+
+        # If the tiledst dataset has any valid data, we read and return it.
+        # Otherwise, we return None.
+        #
+        # To save time, we check the only mask of the first band for
+        # valid data.
+        with memdst.open() as tiledst:
```
Comment (sgillies) [inline mbtiles/__init__.py]: @jdmcbr here's what I was thinking for valid data checking. We've warped the src dataset to a tile-sized GeoTIFF (`tiledst`) that is stored in `memdst`. We can peek at the mask of the first band of `tiledst` to see if there are any valid data pixels. If there are not, we skip the reading of bytes and the worker returns None.

Does that make sense.

File: mbtiles/scripts/cli.py [hunk 4]
```diff
@@ -192,6 +192,10 @@ def mbtiles(ctx, files, output, force_overwrite, title, description,
 
         for tile, contents in pool.imap_unordered(process_tile, tiles):
 
+            if contents is None:
+                logger.info("Tile %r is empty and will be skipped", tile)
+                continue
```
Comment (sgillies) [inline mbtiles/scripts/cli.py]: Empty tiles are logged and skipped here.

File: tests/test_cli.py [hunk 5]
```diff
@@ -76,7 +76,7 @@ def test_export_tiles(tmpdir, data):
     conn = sqlite3.connect(outputfile)
     cur = conn.cursor()
     cur.execute("select * from tiles")
-    assert len(cur.fetchall()) == 6
+    assert len(cur.fetchall()) == 5
```
Comment (sgillies) [inline tests/test_cli.py]: @jdmcbr I had to update the test expectations because one of the tiles was empty and is now skipped :)

**Comment by jdmcbr** [comment]
@sgillies Thanks for taking a shot yourself at this. What I was picturing needing to when you suggested using the memory file was not so different from what I was already trying, so I figured I was missing something, and indeed I was.

Do you a test dataset on which you typically do speed tests? If so, I'd be interested to see your results. I added some tests of this branch on top of what I'd done last night for master vs. #34; I'm out of time now for running the longer tests, but for those I ran, #34 was either as fast or a bit faster than #36. That might just be the particular images I was testing though.

```
master, 50% nodata raster, zoom 10..19
real    0m22.587s
user    0m22.241s
sys     0m1.013s

real    0m22.699s
user    0m22.238s
sys     0m1.060s

#34, 50% nodata raster, zoom 10..19
real    0m13.593s
user    0m21.957s
sys     0m0.913s

real    0m13.679s
user    0m21.750s
sys     0m0.828s


#36, 50% nodata raster, zoom 10..19
real    0m14.211s
user    0m29.032s
sys     0m1.416s

real    0m13.992s
user    0m27.301s
sys     0m1.193s


-------------------------------------------------


master, 0% nodata raster, zoom 10..20
real     0m24.095s
user    0m28.818s
sys     0m1.004s

real    0m24.337s
user    0m28.973s
sys     0m1.088s

#34, 0% nodata raster, zoom 10..20
real    0m23.988s
user    0m29.300s
sys     0m1.008s

real    0m24.026s
user    0m29.923s
sys     0m1.125s

#36, 0% nodata raster, zoom 10..20
real    0m24.633s
user    0m39.681s
sys     0m1.562s

-------------------------------------------------

master, 80% nodata raster, zoom 12..14
real    0m35.410s
user    1m39.214s
sys     0m2.562s


#34, 80% nodata raster, zoom 12..14
real    0m28.367s
user    1m21.301s
sys     0m1.813s


#36, 80% nodata raster, zoom 12..14
real    0m31.256s
user    1m29.678s
sys     0m2.708s


-------------------------------------------------

master, 80% nodata raster, zoom 12..15
real    1m7.456s
user    3m15.152s
sys     0m4.691s


#34, 80% nodata raster, zoom 12..15
real    0m59.370s
user    2m51.536s
sys     0m3.985s


#36, 80% nodata raster, zoom 12..15
real    1m17.884s
user    3m34.508s
sys     0m5.326s
```

File: mbtiles/__init__.py (hunk 2 above)
Comment (sgillies) [reply mbtiles/__init__.py]: @jdmcbr I'm almost certain that the reason #36 is slower than #34 is the extra copy here at line 76.

**Comment by sgillies** [comment]
@jdmcbr I use rasterio's RGB.byte.tif file to benchmark like this:

```
$ command time -l rio mbtiles --zoom-levels 6..10 -f PNG -o /tmp/rgb.mbtiles ~/code/rasterio/tests/data/RGB.byte.tif
        1.63 real         2.81 user         0.25 sys
  63135744  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     53516  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         2  block output operations
         0  messages sent
         0  messages received
         0  signals received
       367  voluntary context switches
      4255  involuntary context switches
```

**Comment by jdmcbr** [comment]
That, surprisingly, had no net impact on the tests I ran. A couple numbers ticked down slightly, a couple others were slightly longer.

**Comment by jdmcbr** [comment]
Current results on the RGB image, zoom 5 to 12 (trying to see time differences more easily). #34 and #36 are identical, both are ~20% faster than master.

```
master, RGB, zoom 5..12
real	0m21.903s
user	0m11.395s
sys	0m0.908s


#34
real	0m16.767s
user	0m11.268s
sys	0m0.810s

real	0m17.102s
user	0m11.643s
sys	0m0.766s


#36
real	0m17.067s
user	0m16.711s
sys	0m0.986s

real	0m16.801s
user	0m17.087s
sys	0m1.120s
```

**Comment by sgillies** [comment]
Closing in favor of #34.

**Pull request closed without merging**